runtime/src/topo/hwloc/topo-hwloc.c

```diff
@@ -281,26 +273,83 @@ void getCPUInfo(void) {
   // get that by counting the parent cores of the accessible PUs.
   //
 
+  // accessible PUs
+
   //
   // We could seemingly use hwloc_topology_get_allowed_cpuset() to get
   // the set of accessible PUs here.  But that seems not to reflect the
   // schedaffinity settings, so use hwloc_get_proc_cpubind() instead.
   //
-  if (hwloc_get_proc_cpubind(topology, getpid(), logAccSet, 0) != 0) {
-#ifdef __APPLE__
-    const int errRecoverable = (errno == ENOSYS); // no cpubind on macOS
-#else
-    const int errRecoverable = 0;
-#endif
-    if (errRecoverable) {
-      hwloc_bitmap_fill(logAccSet);
-    } else {
-      REPORT_ERR_ERRNO(hwloc_get_proc_cpubind(topology, getpid(), logAccSet, 0)
-                       == 0);
-    }
+  if (topoSupport->cpubind->get_proc_cpubind) {
+    int rc = hwloc_get_proc_cpubind(topology, getpid(), logAccSet, 0);
+    CHK_ERR_ERRNO(rc == 0);
+  } else {
+    // assume all PUs are accessible
+    hwloc_bitmap_fill(logAccSet);
   }
   hwloc_bitmap_and(logAccSet, logAccSet,
                    hwloc_topology_get_online_cpuset(topology));
+  numCPUsLogAcc = hwloc_bitmap_weight(logAccSet);
+  CHK_ERR(numCPUsLogAcc > 0);
+
+  //
+  // all PUs
+  //
+  numCPUsLogAll = hwloc_get_nbobjs_by_type(topology, HWLOC_OBJ_PU);
+  CHK_ERR(numCPUsLogAll > 0);
+  //
+  // If some PUs are inaccesssible to us then assume we've been restricted to
+  // certain PUs at a higher level such as by the launcher. If so, use all
+  // accessible PUs. Otherwise, if we can access all PUs but we are sharing
+  // the machine with other locales then the number of locales must equal the
+  // number of sockets and we get our own socket.
+  //
+
+  int numLocalesOnNode = chpl_get_num_locales_on_node();
+  int rank = chpl_get_local_rank();
+  if ((numCPUsLogAcc == numCPUsLogAll) && (numLocalesOnNode > 1) &&
+      (rank != -1)) {
+    int numSockets = hwloc_get_nbobjs_inside_cpuset_by_type(topology,
+                          root->cpuset, HWLOC_OBJ_PACKAGE);
+    if (numSockets != numLocalesOnNode) {
+
+      // If CHPL_RT_LOCALES_PER_NODE is set then the number of locales
+      // must equal the number of sockets.
+      if (chpl_env_rt_get("LOCALES_PER_NODE", NULL) != NULL) {
+        char msg[100];
+        snprintf(msg, sizeof(msg), "The number of locales on the node does "
+                 "not equal the number of sockets (%d != %d).",
+                 numLocalesOnNode, numSockets);
+        chpl_error(msg, 0, 0);
+      } else {
+        // We are oversubscribed.
+        oversubscribed = true;
```

Comment:
I'm having a slow time parsing this code or at least following the logic of when we will or will not auto set oversubscribed. I think before we'd only auto-set it for ofi/ugni since I don't think we set numLocalesOnNode for gasnet. And here I think we'd restricting that so we won't set it if `LOCALES_PER_NODE = processOnNode` (which seems right) but then I'm wondering if this disables any other cases that don't current set rank?